src/critnib/critnib.c

```diff
@@ -410,16 +488,40 @@ void *critnib_remove(struct critnib *c, word key) {
         goto not_found;
     }
 
-    word del =
-        (utils_atomic_increment_u64(&c->remove_count) - 1) % DELETED_LIFE;
+    word del;
+    if (c->cb_free_leaf == NULL) {
+        del = (utils_atomic_increment_u64(&c->remove_count) - 1) % DELETED_LIFE;
+    } else {
+        int i_del = 0;
+        uint64_t ref_count = 0;
+        // We look for a leaf with ref_count == 0.
+        // If we do not find one we take the first one that was checked.
+        // In such case (extremely unlikely) we might not be able to free the value
+        // and a memory leak might happen. Whether it really happened or not
+        // it is verified and logged in alloc_leaf().
+        do {
+            del = (utils_atomic_increment_u64(&c->remove_count) - 1) %
+                  DELETED_LIFE;
+            if (++i_del == (DELETED_LIFE + 1)) {
+                break;
+            }
+
+            k = c->pending_del_leaves[del];
+            if (k) {
+                utils_atomic_load_acquire_u64(&k->ref_count, &ref_count);
+            }
+        } while (k && (ref_count > 0));
+    }
+
```

Comment:
I'm not huge fan of this algoritm. 
- As comment says it can memleak
- Causes that remove count can grow faster, what can reduce performance, as it will cause more retries. 

Can we just use old algorithm then:
- if refcount is non zero we can just remove this node from deleted array(but not free value yet)
- and when refcount == 0, at last unref, we can add this node back to the deleted(free) list.

Reply:
Removed.
Done.